src/system-shutdown.c

```diff
@@ -0,0 +1,153 @@
+        static char *head = NULL;
+        if (!kmsg) {
+                // TODO: figure out why writing to /dev/kmsg doesn't work from here
+                kmsg = stderr;
+                head = "snapd system-shutdown helper: ";
+        }
+
+        va_list va;
+        va_start(va, fmt);
+        fputs(head, kmsg);
+        vfprintf(kmsg, fmt, va);
+        fprintf(kmsg, "\n");
+        va_end(va);
+}
+
+__attribute__ ((noreturn))
+static void die(const char *msg)
+{
+        if (errno == 0) {
+                kmsg("*** %s", msg);
+        } else {
+                kmsg("*** %s: %s", msg, sys_errlist[errno]);
+        }
+        sync();
+        reboot(RB_HALT_SYSTEM);
+        exit(1);
+}
+
+
+// tries to umount all (well, most) things. Returns whether in the last pass it
+// no longer found writable.
+static bool umount_all() {
+        bool did_umount = true;
+        bool had_writable = false;
+
+        for (int i=0; i<10 && did_umount; i++) {
+                struct mountinfo *mounts = parse_mountinfo(NULL);
+                if (!mounts) {
+                        // oh dear
+                        die("unable to get mount info; giving up");
+                }
+                struct mountinfo_entry *cur = first_mountinfo_entry(mounts);
+
+                had_writable = false;
+                did_umount = false;
+                while (cur) {
+                        const char* mount_dir = mountinfo_entry_mount_dir(cur);
+                        const char* mount_src = mountinfo_entry_mount_source(cur);
+                        cur = next_mountinfo_entry(cur);
+
+                        if (strcmp("/", mount_dir) == 0) {
+                                continue;
+                        }
+
+                        if (strcmp("/dev", mount_dir) == 0) {
+                                continue;
+                        }
+
+                        if (strcmp("/proc", mount_dir) == 0) {
+                                continue;
+                        }
+
+                        if (strstr(mount_dir, "writable")) {
+                                had_writable = true;
+                        }
+
+                        if (umount(mount_dir) == 0) {
+                                int fd = open(mount_src, O_RDONLY);
+                                if (fd >=0) {
+                                        ioctl(fd, LOOP_CLR_FD);
+                                        close(fd);
+                                }
+
+                                did_umount = true;
+                        }
+                }
+                cleanup_mountinfo(&mounts);
+        }
+
+        return !had_writable;
+}
+
+
+int main(int argc, char *argv[]) {
+        errno = 0;
+        if (getpid() != 1) {
+                fprintf(stderr, "This is a shutdown helper program; don't call it directly.\n");
+                exit(1);
+        }
+
+        kmsg("started.");
+
+        if (mkdir("/writable", 0755) < 0) {
```

Comment:
Can you document where this program is expecting to execute and what happens if /writable already exists?